crypto/hd/algorithm.go

```diff
@@ -1,96 +1,105 @@
+	//  - eth_secp256k1 (Ethereum)
+	//  - secp256k1 (Tendermint)
+	SupportedAlgorithmsLedger = keyring.SigningAlgoList{EthSecp256k1, hd.Secp256k1}
+)
 
-func DeriveKey(mnemonic, bip39Passphrase, hdPath string, algo keys.SigningAlgo) ([]byte, error) {
-	switch algo {
-	case keys.Secp256k1:
-		return keys.StdDeriveKey(mnemonic, bip39Passphrase, hdPath, algo)
-	case EthSecp256k1:
-		return DeriveSecp256k1(mnemonic, bip39Passphrase, hdPath)
-	default:
-		return nil, errors.Wrap(keys.ErrUnsupportedSigningAlgo, string(algo))
+// EthSecp256k1Option defines a function keys options for the ethereum Secp256k1 curve.
+// It supports eth_secp256k1 and secp256k1 keys for accounts.
+func EthSecp256k1Option() keyring.Option {
+	return func(options *keyring.Options) {
+		options.SupportedAlgos = SupportedAlgorithms
+		options.SupportedAlgosLedger = SupportedAlgorithmsLedger
 	}
+
 }
 
-// EthermintKeygenFunc is the key generation function to generate secp256k1 ToECDSA
-// from ethereum.
-func EthermintKeygenFunc(bz []byte, algo keys.SigningAlgo) (tmcrypto.PrivKey, error) {
-	if algo != EthSecp256k1 {
-		return nil, fmt.Errorf("signing algorithm must be %s, got %s", EthSecp256k1, algo)
-	}
+var (
+	_ keyring.SignatureAlgo = EthSecp256k1
 
-	return ethsecp256k1.PrivKey(bz), nil
+	// EthSecp256k1 uses the Bitcoin secp256k1 ECDSA parameters.
+	EthSecp256k1 = ethSecp256k1Algo{}
+)
+
+type ethSecp256k1Algo struct {
 }
 
-// DeriveSecp256k1 derives and returns the eth_secp256k1 private key for the given mnemonic and HD path.
-func DeriveSecp256k1(mnemonic, bip39Passphrase, path string) ([]byte, error) {
-	hdpath, err := ethaccounts.ParseDerivationPath(path)
-	if err != nil {
-		return nil, err
-	}
+// Name returns eth_secp256k1
+func (s ethSecp256k1Algo) Name() hd.PubKeyType {
+	return EthSecp256k1Type
+}
 
-	seed, err := bip39.NewSeedWithErrorChecking(mnemonic, bip39Passphrase)
-	if err != nil {
-		return nil, err
-	}
+// Derive derives and returns the eth_secp256k1 private key for the given mnemonic and HD path.
+func (s ethSecp256k1Algo) Derive() hd.DeriveFn {
+	return func(mnemonic string, bip39Passphrase, path string) ([]byte, error) {
+		hdpath, err := ethaccounts.ParseDerivationPath(path)
+		if err != nil {
+			return nil, err
+		}
 
-	masterKey, err := hdkeychain.NewMaster(seed, &chaincfg.MainNetParams)
-	if err != nil {
-		return nil, err
-	}
+		seed, err := bip39.NewSeedWithErrorChecking(mnemonic, bip39Passphrase)
+		if err != nil {
+			return nil, err
+		}
 
-	key := masterKey
-	for _, n := range hdpath {
-		key, err = key.Child(n)
+		masterKey, err := hdkeychain.NewMaster(seed, &chaincfg.MainNetParams)
 		if err != nil {
 			return nil, err
 		}
-	}
 
-	privateKey, err := key.ECPrivKey()
-	if err != nil {
-		return nil, err
+		key := masterKey
+		for _, n := range hdpath {
+			key, err = key.Child(n)
+			if err != nil {
+				return nil, err
+			}
+		}
```

Comment:
this will go through the derivation paths array and only output the last one. is this intended or am i not understanding the hdpath correctly here